### ISSUE TITLE
React does not recognize the XX prop on a DOM element

### ISSUE THREAD

**Issue by devrnt** [cristianbote/goober#203]
**Description**
React console error whenever passing props to a goober styled component:
`
React does not recognize the XX prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase XX instead. If you accidentally passed it from a parent component, remove it from the DOM element.
`

**When happening**
Passing through some props to a goober styled component (see codesandbox)

**Possible solution?**
See `transient props` from `styled-components`

**Codesandbox**
See the console logs
https://codesandbox.io/s/vigorous-gates-0o2ge?file=/src/App.js

**Comment by cristianbote** [comment]
Gosh, I really do not want to have to use `$<name>` props naming 🤕  looking into it

**Comment by devrnt** [comment]
Hmm, I currently see no other option than keeping a list of "default" props (don't know if this is desirable).

I made a fork to look at a solution with a `$` prefix and it seems to work. I'll keep this PR on idle until this ticket gets any update

**Comment by jayalfredprufrock** [comment]
Landed here after using the cxs project which had similar goals but is no longer actively maintained. That library solved the problem by automatically stripping properties defined on propTypes (https://github.com/cxs-css/cxs/blob/eb934203a17e07aec63e971c3b87068e911b5b6e/src/component.js#L11) In combination with a babel transform that generates the propTypes from typescript typings, the developer experience was pretty darn good, with the obvious downside of having to ship propTypes in production.

Some might find the small increase in bundle size worth it to not have to prefix every style prop, especially when integrating with something like styled-system. I would also welcome the ability to pass off the burden to the developer via a shouldForwardProp function like what emotion exposes. Or perhaps a babel plugin could be used to generate an optimal black/white list of props so the library wouldn't have to ship with an exhaustive list...idk just thinking out loud. I just know that I've seen the magical prefix approach pitched in numerous style libraries and it always seemed like more people were against it than for it...then again I could be biased on that front :) Either way, this library is looking like an attractive alternative to cxs and I plan on giving it a spin soon!

**Comment by cristianbote** [comment]
Hey @jayalfredprufrock thank you so much for weighing in! Much appreciate it!

After reading your argument and looking into emotion's `shouldForwardProp` functionality and I am really digging it! `goober` currently supports `forwardRef` via the second argument, so the second argument is taken. But, what if you could pass a prop filter to `setup` instead? 🤔 

```js
setup(h, prefix, theme, shouldForwardRef);

function shouldForwardRef(propertyName) {
  return propertyName !== 'size';
}
```

What are your thoughts @devrnt?

**Comment by devrnt** [comment]
@cristianbote @jayalfredprufrock Feels like the perfect solution to me, I opened a new PR since I accidentally removed my repo (damn me):  #227

**Comment by smiley-uriux** [comment]
Love it!

**Comment by jayalfredprufrock** [comment]
Love it!

**Comment by cristianbote** [comment]
@devrnt thank you for doing the weight lifting in here! I've added a comment on the latest changes: https://github.com/cristianbote/goober/pull/227/files#r552475713.

Basically, the solution with the least impact on size it's if instead of iterating over _each_ of the prop, we let the user do it for us:
```js
setup(pragma, prefixer, theme, shouldForwardProps);

function shouldForwardProps(props) {
  const allowed = ['foo', 'bar'];

  for (let prop in props) {
    if (!allowed.contains(prop)) {
      delete props[prop];
    }
  }
}
```

**Comment by atersolis** [comment]
I like the idea of having a shouldForwardProps function as a parameter for setup.
But I think it would be great if goober could also provide a function in a separate package like "goober/should-forward-props" or something.

I took a look at how they do this in 'styled-components'. It turns out that they use @emotion/is-prop-valid.
The core logic is pretty straightforward [you can read it here](https://github.com/emotion-js/emotion/blob/master/packages/is-prop-valid/src/index.js).
According to bundlephobia it's 2.1kB Minified + Gzipped, because they ship a list of the valid props.
It's a small library but it's still twice the size of goober.

So maybe we could use the DOM api to check if a prop is valid ?
I did a little research but I didn't find anything that would return a correct result...

However if you are willing to make some compromise, then this might do the job :
```javascript
propName in document.createElement(tagName)
```
The problem with this is that props like innerHTML, cloneNode, etc... would be forwarded
So this is not a "correct" function, but it would work in many cases.
If we combine this with a regex, we get something that covers most of the practical use cases I think:

```javascript
const shouldForwardProps = memoize((propName, tagName) =>
   /^(on|data-|aria-)/.test(propName) || (propName in document.createElement(tagName))
)
```

What do you think ?

**Comment by cristianbote** [comment]
@atersolis thank you so much for weighing in on this! 😄 happy to see the community participate in _any_ discussion!

Yeah a separate package sounds ideal as it could be extended or not, depending on what the user wants to achieve.

The logic based on browser detection, it's really odd when you take into consideration the server-side part. Since goober supports server environments that would mean either have the user load a virtual dom or dom-like apis.

I believe we could for now, support a function that would receive the props and that function should handle the cleanup.

**Comment by cristianbote** [comment]
@devrnt published `2.0.20` with `forwardProps` 🥳  thank you so much for landing this! This was all your doing! So happy to see it in `goober`.
